src/native_background.ts

```diff
@@ -161,16 +224,55 @@ export async function mkdir(dir: string, exist_ok: boolean) {
 export async function temp(content: string) {
     return sendNativeMsg("temp", { content })
 }
+
 export async function run(command: string) {
     let msg = await sendNativeMsg("run", { command })
     logger.info(msg)
     return msg
 }
 
+export async function getenv(variable: string) {
+    let v = await getNativeMessengerVersion()
+    if (!nativegate("0.1.2", false)) {
+        throw `Error: getenv needs native messenger v>=0.1.2. Current: ${v}`
+    }
+    return (await sendNativeMsg("env", { var: variable })).content
+}
+
+/** This returns the commandline that was used to start firefox.
+ You'll get both firefox binary (not necessarily an absolute path) and flags */
+export async function ffargs(): Promise<string[]> {
+    // Using ' and + rather that ` because we don't want newlines
+    let output = await sendNativeMsg("eval", {
+        command:
+            "handleMessage(" +
+            '{"cmd": "run", "command": "ps -p " + str(os.getppid()) + " -o' +
+            'args="})["content"]',
+    })
+    return output.content.trim().split(" ")
+}
+
 export async function getProfileDir() {
+    // First, see if we can get the profile from the arguments that were given
+    // to Firefox
+    let args = await ffargs()
+
+    // --profile <path>: Start with profile at <path>
+    let prof = args.indexOf("--profile")
+    if (prof >= 0) return args[prof + 1]
+
+    // -P <profile>: Start with <profile>
+    // -P          : Start with profile manager
+    let profileName = "*"
+    prof = args.indexOf("-P")
+    // args.length -1 because we need to make sure -P was given a value
+    if (prof >= 0 && prof < args.length - 1) profileName = args[prof + 1]
+
     // Find active profile directory automatically by seeing where the lock exists
-    let hacky_profile_finder =
-        "find ../../../.mozilla/firefox -maxdepth 2 -name lock"
+    // We can't use a relative path because ~/.local (the directory where the
+    // native messenger currently sits) might actually be a symlink to another dir
+    let home = await getenv("HOME")
```

Comment:
You should put a try...catch around this little bit and fall back to a ../../../ path, otherwise we're breaking things for users of <0.1.2 needlessly.